src/envs/playroom.py

```diff
@@ -0,0 +1,631 @@
+            _initiable=self._PutOnTable_initiable,
+            _terminal=self._PutOnTable_terminal)
+        # Still need to add new options
+        # Objects
+        self._robot = Object("robby", self._robot_type)
+        self._door1 = Object("door1", self._door_type)
+        self._door2 = Object("door2", self._door_type)
+        self._door3 = Object("door3", self._door_type)
+        self._door4 = Object("door4", self._door_type)
+        self._door5 = Object("door5", self._door_type)
+        self._door6 = Object("door6", self._door_type)
+        self._doors = (self._door1, self._door2, self._door3,
+                       self._door4, self._door5, self._door6)
+        self._dial = Object("dial", self._dial_type)
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        x, y, z, _, fingers = action.arr
+        # Update robot position
+        if not self._is_valid_loc(action):
+            return state.copy()
+        if self._robot_can_move(state, action):
+            state = self._transition_move(state, action)
+
+        x = state.get(self._robot, "pose_x")
+        y = state.get(self._robot, "pose_y")
+        # Interact with blocks
+        if (self.table_x_lb < x < self.table_x_ub) \
+            and (self.table_y_lb < y < self.table_y_ub) \
+            and self._robot_is_facing_table(action):
+            if fingers < 0.5:
+                return self._transition_pick(state, action)
+            if z < self.table_height + CFG.playroom_block_size:
+                return self._transition_putontable(state, action)
+            return self._transition_stack(state, action)
+        # Interact with some door
+        if any(self._NextToDoor_holds(state, (self._robot, door))
+               for door in self._doors):
+            door = self._get_door_next_to(state)
+            door_x = state.get(door, "pose_x")
+            door_y = state.get(door, "pose_y")
+            if (door_x-self.door_tol < x < door_x+self.door_tol) \
+                and (door_y-self.door_tol < y < door_y+self.door_tol) \
+                and (self.door_button_z-self.door_tol < z
+                        < self.door_button_z+self.door_tol) \
+                and fingers >= self.open_fingers \
+                and self._robot_is_facing_door(state, action, door):
+                return self._transition_door(state, door)
+        # Interact with dial
+        dial_x = state.get(self._dial, "pose_x")
+        dial_y = state.get(self._dial, "pose_y")
+        if (dial_x-self.dial_button_tol < x < dial_x+self.dial_button_tol) \
+            and (dial_y-self.dial_button_tol < y
+                    < dial_y+self.dial_button_tol) \
+            and fingers >= self.open_fingers \
+            and self._robot_is_facing_dial(state, action):
+            return self._transition_dial(state)
+
+        return state.copy()
+
+    def _transition_pick(self, state: State, action: Action) -> State:
```

Comment:
for this function and the others, I'm wondering if they're identical to the parent class? if so, remove; if not, please add a comment explaining the difference, or, if it involves a small change that would make sense in the parent class too, make the change there.

Reply:
the only difference is the action space...I could refactor it by having the functions take in, e.g., `x`, `y`, `z`, and `fingers` instead of `action`? for a couple of the other functions, the only difference is `CFG.blocks_block_size` vs. `CFG.playroom_block_size`, so could I change that to a class variable `block_size` instead?

Reply:
yes, both of those changes sound great. thank you!!